library/Zend/Form/Element/Collection.php

```diff
@@ -221,6 +222,10 @@ public function populateValues($data)
                 ));
             }
 
+            $removed[] = $name;
+        }
+
+        foreach ($removed as $name) {
```

Comment:
I've tried a few different approaches at this time. The problem is that doing removal while inside the loop leads to inconsistent results; the underlying iterator is not updated in place correctly. As such, the first pass is to determine which we can remove, and then we remove each.